eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs

```diff
@@ -109,6 +125,53 @@ pattern Data p <-
 getPlutusData :: Data era -> Plutus.Data
 getPlutusData (DataConstr (Memo d _)) = d
 
+-- | Inlined data must be stored in the most compact form because it contributes
+-- to the memory overhead of the ledger state. Constructor is intentionally not
+-- exported, in order to prevent invalid creation of data from arbitrary binary
+-- data. Use `makeInlineData` for smart construction.
+newtype InlineData era = InlineData ShortByteString
```

Comment:
In order to guarantee type safety `Data` and `InlineData` have to live in the same module (note both of their constructors aren't exported). I think it is more important than placing types into eras they were introduced in.

However, if it is the name that bothers you, I can rename it to `BinaryData` and then it will not be tight into Babbage in any way.

Reply:
I sometimes wish Haskell had a concept of "friends", a la C++. But yeah, I wonder whether `BinaryData` might be a better name, just to avoid confusion.

Reply:
Isn't the usual trick to make `Cardano.Ledger.Alonzo.Data.Internal`, and re-export in `Cardano.Ledger.Alonzo.Data`? I'm happy either way.

Reply:
Using `Internal` modules is one of the common ways to deal with this, but that doesn't give you the same guarantees as the approach taken in this PR, because `Internal` modules simply relay responsibility to all programmers, sort of: "if you want to shoot yourself in the foot this is how you do it". While not exporting constructors at all, and instead providing smart constructors (a.k.a `Data`  and `makeInlineData` in this case) will give hard guarantees that undeserializable `InlineDatum` cannot be created outside the `Data` module.

I've renamed it to `BinaryData`. This will avoid any future confusion and will provide a good insight into what that type means.

Reply:
I guess the nice part about "friends" is that you can pick exactly who all gets to shoot themselves in the foot. :)